language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/MakeFunctionIsolatedCodeAction.java

```diff
@@ -0,0 +1,134 @@
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.ballerinalang.langserver.commons.CodeActionContext;
+import org.ballerinalang.langserver.commons.codeaction.spi.DiagBasedPositionDetails;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.CodeActionKind;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextEdit;
+
+import java.net.URI;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Code Action to make function isolate.
+ *
+ * @since 2201.1.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class MakeFunctionIsolatedCodeAction extends AbstractCodeActionProvider {
+    public static final String NAME = "Make Function Isolated";
+    public static final String DIAGNOSTIC_CODE_BCE3946 = "BCE3946";
+    public static final String DIAGNOSTIC_CODE_BCE3947 = "BCE3947";
+
+    @Override
+    public List<CodeAction> getDiagBasedCodeActions(Diagnostic diagnostic,
+                                                    DiagBasedPositionDetails positionDetails,
+                                                    CodeActionContext context) {
+        if (!DIAGNOSTIC_CODE_BCE3946.equals(diagnostic.diagnosticInfo().code()) 
+                && !DIAGNOSTIC_CODE_BCE3947.equals(diagnostic.diagnosticInfo().code()) 
+                || context.currentSyntaxTree().isEmpty() || context.currentSemanticModel().isEmpty()) {
+            return Collections.emptyList();
+        }
+        Range diagnosticRange = CommonUtil.toRange(diagnostic.location().lineRange());
+        NonTerminalNode nonTerminalNode = CommonUtil.findNode(diagnosticRange, context.currentSyntaxTree().get());
+        
+        if (nonTerminalNode.kind() == SyntaxKind.EXPLICIT_ANONYMOUS_FUNCTION_EXPRESSION) {
+            return getExplicitAnonFuncExpressionCodeAction((ExplicitAnonymousFunctionExpressionNode) nonTerminalNode, 
+                    context);
+        }
+        Optional<Symbol> symbol = context.currentSemanticModel().get().symbol(nonTerminalNode);
+        if (symbol.isEmpty() || symbol.get().getModule().isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        Optional<Project> project = context.workspace().project(context.filePath());
+        if (project.isEmpty()) {
+            return Collections.emptyList();
+        }
+        Optional<Path> filePath = CommonUtil.getFilePathForSymbol(symbol.get(), project.get(), context);
+        if (filePath.isEmpty() || context.workspace().syntaxTree(filePath.get()).isEmpty()) {
+            return Collections.emptyList();
+        }
+        URI uri = filePath.get().toUri();
+
+        Optional<NonTerminalNode> node = CommonUtil.findNode(symbol.get(), 
+                context.workspace().syntaxTree(filePath.get()).get());
+        if (node.isEmpty()) {
```

Comment:
You can check if the node kind is FUNCTION_DEFINITION here

Reply:
What if the node.isEmpty() is True while node kind is FUNCTION_DEFINITION?

Reply:
You can check for both and pass a FunctionDefintionNode to the getStartPosition() method. Instead of NonTerminalNode. Doing so we can get rid of extra checks inside the getStartPosition() method.

Reply:
You can check for both and pass a FunctionDefintionNode to the getStartPosition() method. Instead of NonTerminalNode. Doing so we can get rid of extra checks inside the getStartPosition() method.

Reply:
if (node.isEmpty() || node.kind() != SyntaxKind.FunctionDefinition)